feat: add Learning-Based Accel/Brake Map Calibrator (#231)
autowarefoundation/autoware_tools: vehicle/learning_based_vehicle_calibration/README.md

You can visualize the collection process from the terminal.

![data_collection1](./imgs/data_collection1.png)

Otherwise, we built some custom messages for representing the progress that are being published on the following topics:

```bash
/scenarios_collection_longitudinal_throttling

/scenarios_collection_longitudinal_braking
```

Once you have collected the data, in order to train and build your black box models, for both throttling and braking scenarios, launch:

```bash
ros2 launch learning_based_vehicle_calibration neural_network_launch.py
```

You will obtain the acceleration and braking calibration maps and visualize how the neural network fits the data:

![NN_throttle](./imgs/NN_throttle.png)

## Overview

Here we present the software structure, data collection, data preprocessing and neural network training and visualization about the end-to-end calibration of a vehicle, in two different scenarios:

- **Normal driving conditions (longitudinal dynamics);**
- **Parking conditions (steering dynamics).**

## Input Data Software

Launch Autoware as follows:

```bash

> ./autoware.sh

Reviewer: Where is `autoware.sh` located?

Author: @xmfcx actually this repo was initially designed for Pix Moving, in the README I just added that reminder to my colleagues not to forget to launch Autoware before the calibrator.

I forgot to remove that line before submitting this PR, sorry

Author: Hello @xmfcx,

I noticed that the existing accel/brake map calibrator is available in the autoware_universe repository ( https://github.com/autowarefoundation/autoware_universe/tree/main/vehicle/autoware_accel_brake_map_calibrator ). Our new learning‑based calibrator builds on this work by improving both usability and performance. Would you recommend integrating it into autoware_universe rather than autoware_tools? Thank you

Reviewer: Hi @CristianGariboldi ,

I've also consulted with @mitsudome-r and in the future we will move that calibrator to this repository as well.

It's better to have it here because this is a tool that won't be used during the autonomous driving session.

Author: @xmfcx Crystal clear, thank you very much for your feedback!
I am of course available if you need further clarifications and willing to help.

Since we want to generate a calibration table which takes throttle/brake and speed as inputs, and outputs acceleration, we can follow the same structure for the neural network.

So the fully connected neural network will have 2 nodes in the input layer and 1 node in the output layer.

Regarding the hidden layers, we can't know in advance how many layers and parameters we need in order to fit our data in the best way. So, based on the performance we obtain according to some metrics such as MSE, MAE, RMSE and R2, we can modify the structure and the number of parameters to be trained in order to optimize our metrics.

Now we can split the dataset into training set (80%) and testing set (20%) and we are ready for the training stage.

We chose the ReLu function as the activation function, mean square error as the cost function and Adam optimizer, all in Pytorch.

To start the training of the throttling and braking models and visualize their shapes and metrics, we can launch:

```bash
ros2 launch learning_based_vehicle_calibration neural_network_launch.py
```

![NN_throttle](./imgs/NN_throttle.png)

![NN_brake](./imgs/NN_brake.png)

> When you launch these scripts, the training stage begins and when it is done, the software automatically saves the trained model in a zip file.
> 
> This way, when you want to use the neural network model, you can avoid to train it every time from scratch, but you can just train it once and then load it, according to the following scripts:
> 
> These scripts will also create the acceleration and braking offline tables.

Reviewer: ```suggestion
When you launch these scripts, the training stage begins and when it is done, the software generates the acceleration and braking offline tables: `accel_map.csv` and `brake_map.csv` in your working directory.
```

I did not observe a zip or model file.
